### PR TITLE
issue-434:Convert Pytest-CVP API access test case to Vane

### PR DESCRIPTION
**Pull request by rohitsinghgslab**
# Please include a summary of the changes

Convert Pytest-CVP API access test case to Vane as per the vane style guide along with that updated following files.
test_definition.yaml: Updated test def with testcase NRFU5.1
test_definition.yaml.j2  : Updated J2 with testcase NRFU5.1

# Any specific logic/part of code you need extra attention on

Verified ACLs are applied to each VRF the API is enabled on.

# Include the Issue number and link
#434 

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [ ] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [x] Other (please specify)

# Effort required on reviewers end

- - [ ] Easy
- - [x] Medium
- - [ ] Hard 

# How Has This Been Tested?
**Note: Tested with using hardcoded output**
Unit tests and report :
**Reports:**
Pass case:All VRFs that the API is active are ACL configured.(J2)
[report_2309201219.docx](https://github.com/aristanetworks/vane/files/12669735/report_2309201219.docx)

Fail case: All VRFs that the API is active are not ACL configured.(J2)
[report_2309201214.docx](https://github.com/aristanetworks/vane/files/12669743/report_2309201214.docx)

Assert VRFs: Vrfs details are not found in output(Hard coded, as command is giving vrfs on every device.).(J2)
[report_2309201156.docx](https://github.com/aristanetworks/vane/files/12669745/report_2309201156.docx)

Assert ACLs: ACL details for APIs is not found in output.(J2)
[report_2309201149.docx](https://github.com/aristanetworks/vane/files/12669748/report_2309201149.docx)

HTML Reports(with J2):
[PS_434_HTML.zip](https://github.com/aristanetworks/vane/files/12669766/PS_434_HTML.zip)

Pass case:All VRFs that the API is active are ACL configured.(Without J2)
[Passed_without_J2.zip](https://github.com/aristanetworks/vane/files/12669810/Passed_without_J2.zip)

## New feature
    
# CI pipeline result

- - [ ] Pass
- - [ ] Fail
  
  If it fails, explain the reason and whether or not we should ignore the failure
  
# Verify Documentation Update

    
# Additional comments

### CODE REVIEW

File: nrfu_tests/test_acls_api_access_list.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Copyright (c) 2023 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+"""
+Testcase for verification of ACLs are applied to each VRF the API is enabled on.
```
Comment (AbhijeetGSLAB) [inline nrfu_tests/test_acls_api_access_list.py:5]: Change the file name to test_security_rp_api_access_list.py

File: nrfu_tests/test_acls_api_access_list.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                        tops.actual_output["vrf_details"].update({vrf: "Configured"})
+                    if not acl_found:
+                        tops.actual_output["vrf_details"].update({vrf: "Not Configured"})
+
+            # Forming output message if test result is failed.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                non_configured_vrfs = []
+                for vrf_name, vrf_status in tops.expected_output["vrf_details"].items():
+                    actual_vrf_status = tops.actual_output["vrf_details"].get(vrf_name)
```
Comment (AbhijeetGSLAB) [inline nrfu_tests/test_acls_api_access_list.py:94]: expected_vrf_status

File: nrfu_tests/test_acls_api_access_list.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    if vrf_status != tops.actual_output["vrf_details"].get(vrf_name):
+                        non_configured_vrfs.append(f"{vrf_name} - {actual_vrf_status}")
+                vrf_configured_status = "\n".join(non_configured_vrfs)
+                tops.output_msg += (
+                    f"Following Vrfs are not ACL configured:\n{vrf_configured_status}."
```
Comment (AbhijeetGSLAB) [inline nrfu_tests/test_acls_api_access_list.py:99]:
```suggestion
                    f"Following vrfs are not ACL configured:\n{vrf_configured_status}."
```

File: nrfu_tests/test_acls_api_access_list.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                acl_found = False
+                for acl in api_acls:
+                    tops.expected_output["vrf_details"].update({vrf: "Configured"})
+                    if vrf in acl["activeVrfs"]:
+                        acl_found = True
+                        tops.actual_output["vrf_details"].update({vrf: "Configured"})
+                    if not acl_found:
+                        tops.actual_output["vrf_details"].update({vrf: "Not Configured"})
+
+            # Forming output message if test result is failed.
```
Comment (AbhijeetGSLAB) [inline nrfu_tests/test_acls_api_access_list.py:89]: image

Can we just add the vrf names that are not configured. As in the msg we are stating that its not configured and the same is added again below.

File: nrfu_tests/test_acls_api_access_list.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        self.output = ""
+
+        # Forming output message if test result is passed.
+        tops.output_msg = "All VRFs that the API is active are ACL configured."
+
+        try:
+            """
+            TS: Running `show management api http-commands` command, collecting VRFs and
+            running `show management api http-commands ip access-list summary` command,
+            collecting ACLs. Also, verifying VRFs are acl configured.
```
Comment (AbhijeetGSLAB) [inline nrfu_tests/test_acls_api_access_list.py:48]:
```suggestion
            TS: Running `show management api http-commands` and
            `show management api http-commands ip access-list summary` commands
            and verifying that all the VRFs are ACL configured.
```

File: nrfu_tests/test_acls_api_access_list.py (hunk 2 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_acls_api_access_list.py:94]: removed this, as in suggested failure msg it is not required.

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+            # Collecting actual and expected output.
+            for vrf in api_vrfs:
+                acl_found = False
+                for acl in api_acls:
+                    tops.expected_output["vrf_details"].update({vrf: "Configured"})
+                    if vrf in acl["activeVrfs"]:
+                        acl_found = True
+                        tops.actual_output["vrf_details"].update({vrf: "Configured"})
+                    if not acl_found:
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_security_rp_api_access_list.py:86]: check this condition

**Review by AbhijeetGSLAB — APPROVED** [review]

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        tops.actual_output = {"vrf_details": {}}
+        self.output = ""
+
+        # Forming output message if test result is passed.
+        tops.output_msg = "All VRFs that the API is active are ACL configured."
+
+        try:
+            """
+            TS: Running `show management api http-commands` and
+            `show management api http-commands ip access-list summary` commands
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_security_rp_api_access_list.py:47]: separate out the TS for each show command

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            # Forming output message if test result is failed.
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = "\n"
+                non_configured_vrfs = []
+                for vrf_name, vrf_status in tops.expected_output["vrf_details"].items():
+                    if vrf_status != tops.actual_output["vrf_details"].get(vrf_name):
+                        non_configured_vrfs.append(vrf_name)
+                vrf_configured_status = ", ".join(non_configured_vrfs)
+                tops.output_msg += (
+                    f"Following vrfs are not ACL configured: {vrf_configured_status}."
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_security_rp_api_access_list.py:98]: Take the reference of PR 474 for TS/observation

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        tops.output_msg = "All VRFs that the API is active are ACL configured."
+
+        try:
+            """
+            TS: Running `show management api http-commands` and
+            `show management api http-commands ip access-list summary` commands
+            and verifying that all the VRFs are ACL configured.
+            """
+            output_api_vrfs = dut["output"][tops.show_cmds[tops.dut_name][0]]["json"]
+            output_api_acls = dut["output"][tops.show_cmds[tops.dut_name][1]]["json"]
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_security_rp_api_access_list.py:51]: dut output should match with observation. Dont use hardcode output
image

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            # Collecting actual and expected output.
+            for vrf in api_vrfs:
+                acl_found = False
+                for acl in api_acls:
+                    tops.expected_output["vrf_details"].update({vrf: "Configured"})
+                    if vrf in acl["activeVrfs"]:
+                        acl_found = True
+                        tops.actual_output["vrf_details"].update({vrf: "Configured"})
+                    if not acl_found:
+                        tops.actual_output["vrf_details"].update({vrf: "Not Configured"})
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_security_rp_api_access_list.py:87]: Please update your actual and expected output. configured/not configured is giving proper details. Take the reference of PR 474
image

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+@pytest.mark.nrfu_test
+@pytest.mark.security
+class AclsApiAccessTests:
+    """
+    Testcase for verification of ACLs are applied to each VRF the API is enabled on.
+    """
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_acls_api_vrfs_enabled"]["duts"]
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_security_rp_api_access_list.py:25]: Why you are not running the testcase on BOFA lab?
image

**Review by MaheshGSLAB — CHANGES_REQUESTED** [review]

**Review by MaheshGSLAB — CHANGES_REQUESTED** [review]

**Review by MaheshGSLAB — APPROVED** [review]

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            tests_definitions(dict): test suite and test case parameters.
+        """
+        tops = tests_tools.TestOps(tests_definitions, TEST_SUITE, dut)
+        tops.expected_output = {"vrfs": {}}
+        tops.actual_output = {"vrfs": {}}
+        self.output = ""
+        show_cmds = tops.show_cmds[tops.dut_name]
+
+        # Forming output message if the test result is passed.
+        tops.output_msg = "All VRFs that the API is active on, an ACL is configured."
```
Comment (rewati-arista) [inline nrfu_tests/test_security_rp_api_access_list.py:43]: Can we rephrase this to "ACL is configured on all VRF's which have API enabled". Across the changed files the same sentence has been conveyed in many ways already, lets try and keep the phrasing simple and consistent 😅

File: nrfu_tests/test_security_rp_api_access_list.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                f" is:\n{output_api_acls}\n"
+            )
+            self.output = f"Output of {show_cmds[1]} command is:\n{output_api_acls}\n"
+
+            # Collecting VRFs and ACLs from output.
+            api_vrfs = output_api_vrfs.get("vrfs")
+            api_acls = output_api_acls.get("ipAclList").get("aclList")
+            assert api_vrfs, "Vrfs details are not found in the output."
+
+            assert api_acls, "ACL details for APIs is not found in the output."
```
Comment (rewati-arista) [inline nrfu_tests/test_security_rp_api_access_list.py:73]:
```suggestion
            assert api_acls, "ACL details for APIs are not found in the output."
```

File: nrfu_tests/test_definition.yaml [hunk 14]
```diff
@@ -135,13 +135,16 @@
...
+    - name: test_acls_api_vrfs_enabled
+      description: Test case to verify that ACL is configured for each VRF on which API is enabled.
       test_id: NRFU5.1
-      show_cmd: null
-      expected_output: null
+      show_cmds:
+        - show management api http-commands
+        - show management api http-commands ip access-list summary
+      expected_output: null # Forming expected output in test case.
+      test_criteria: All VRFs that the API is active on should have an ACL to be configured. # Setting test case’s pass/fail criteria for reporting
```
Comment (rewati-arista) [inline nrfu_tests/test_definition.yaml:145]: lets keep phrasing consistent between description and criteria
```suggestion
      test_criteria: ACL should be configured on all VRFs on which API is enabled. # Setting test case’s pass/fail criteria for reporting
```

File: nrfu_tests/test_definition.yaml.j2 [hunk 15]
```diff
@@ -120,11 +120,14 @@
...
-      description:
+    - name: test_acls_api_vrfs_enabled
+      description: Test case to verify that ACL is configured for each VRF on which API is enabled.
       test_id: NRFU5.1
-      show_cmd: null
+      show_cmds:
+        - show management api http-commands
+        - show management api http-commands ip access-list summary
       expected_output: null
+      test_criteria: All VRFs that the API is active on should have an ACL to be configured.
```
Comment (rewati-arista) [inline nrfu_tests/test_definition.yaml.j2:130]: same suggestion as above

**Review by rewati-arista — CHANGES_REQUESTED** [review]
Minor rephrasing suggestions, rest looks good

File: nrfu_tests/test_definition.yaml.j2 (hunk 15 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_definition.yaml.j2]: Done

File: nrfu_tests/test_security_rp_api_access_list.py (hunk 12 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_security_rp_api_access_list.py]: True, changed it.

File: nrfu_tests/test_definition.yaml (hunk 14 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_definition.yaml]: Done

**Review by rewati-arista — APPROVED** [review]
LGTM
